src/udev/udev-builtin-blkid.c

```
@@ -129,101 +131,163 @@ static int find_gpt_root(UdevEvent *event, blkid_probe pr, const char *loop_back
-                 * systematic behaviour. */
-                if (!STRPTR_IN_SET(loop_backing_fname, "rootdisk", "rootdisk.raw")) {
-                        log_device_debug(dev, "Device is not a loopback block device with reference string 'root', not considering block device as default root block device.");
+                        /* OK, we have now sufficiently identified this device as the right root "whole" device,
+                         * hence no need to bother with searching for ESP/XBOOTLDR */
+                        need_esp_or_xbootldr = false;
+                } else
+                        /* We now know the the ESP/xbootldr UUID, but we cannot be sure yet it's on this block
+                         * device, hence look for it among partitions now */
+                        need_esp_or_xbootldr = true;
+        } else {
+                /* On the main system look at the *current* root instead */
+
+                r = blockdev_get_root(LOG_DEBUG, &root_devno);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Unable to determine current root block device, skipping gpt-auto probing: %m");
+                        return 0;
+                }
+                if (r == 0) {
+                        log_device_debug(dev, "Root block device not backed by a (single) whole block device, skipping gpt-auto probing.");
                         return 0;
                 }
 
-                /* OK, we have now sufficiently identified this device as the right root "whole" device,
-                 * hence no need to bother with searching for ESP/XBOOTLDR */
+                dev_t whole_devno;
+                r = block_get_whole_disk(root_devno, &whole_devno);
+                if (r < 0)
+                        log_device_debug_errno(dev, r, "Unable to find whole block device for root block device, ignoring: %m");
```

If the error is ignored, then we need to skip the below check, otherwise, uninitialized value in `whole_devno` will be used in the below.